Fix93: Improve Container Integration

PhMueller opened the pull request:
This PR includes: 
* On some clusters, the HPOBench was not able to mount the necessary directories into the container. As consequence, the container was not able to read the configuration file. But since the container does not depend on the configuration file, it does not read the hpobenchrc file anymore. (see #93)
* The local socket/cache/data are now linked to fix directories in the container. (var/lib/hpobench/x) via the --env parameter. 
* I have removed the global data directory from the hpobench, because it seems to be legacy and I can't see a clear use-case for it anymore.

The next steps are:
* Add a note to the readme that the user needs to have read-write rights to the 3 directories cache, socket, data dir. 
* Upload some updated containers and try it on the Nemo cluster

PhMueller committed:
[Skip CI] [WIP] Improve Container Integration

PhMueller committed:
FIX environmental variables

Moved the environment vars.
I have used a feature (passing via --env) which is not available in singularity v3.5, but comes with 3.6.

KEggensperger commented on hpobench/config.py:
```diff
@@ -41,18 +33,33 @@ def __init__(self):
+        self.config_base_dir = Path(os.environ.get('XDG_CONFIG_HOME', '~/.config/hpobench')).expanduser().absolute()
+
+        # The path to the hpobenchrc file. This file stores the settings defined by the user
+        self.config_file = (self.config_base_dir / '.hpobenchrc').expanduser().absolute()
```
expanduser() and absolut() are redundant

KEggensperger commented on hpobench/config.py:
```diff
@@ -41,18 +33,33 @@ def __init__(self):
+        # Constant paths in the container
+        self._cache_dir_container = '/var/lib/hpobench/cache'
+        self._data_dir_container = '/var/lib/hpobench/data'
+        self._socket_dir_container = '/tmp/hpobench_socket/'
```
Shouldn't make a difference, but only one path ends with "/"

KEggensperger commented on hpobench/config.py:
```diff
@@ -41,18 +33,33 @@ def __init__(self):
+
+        # Constant paths in the container
+        self._cache_dir_container = '/var/lib/hpobench/cache'
+        self._data_dir_container = '/var/lib/hpobench/data'
```
Does this mean that when using containers, datasets will be downloaded with every new instantiation of the container?

KEggensperger commented on hpobench/container/client_abstract_benchmark.py:
```diff
@@ -62,14 +62,15 @@ class AbstractBenchmarkClient(metaclass=abc.ABCMeta):
 
     def _setup(self, benchmark_name: str, container_name: str, container_tag: str,
-               container_source: Optional[str] = None,
+               container_source: Optional[str] = None, env_str: Optional[str] = '', bind_str: Optional[str] = '',
```
Since args for `__init__()` and `_setup()` are the same, could you also order them in the same way?

PhMueller committed:
Increase Version Number for each Container

PhMueller committed:
Equalize client abstract benchmark function calls.

PhMueller committed:
Allow changing the socket dir via environment variable

PhMueller committed:
Paramnet is now 0.0.4 because of another PR.

PhMueller committed:
Merge branch 'development' into FIX93_ContainerConfig

PhMueller committed:
Fix in Socket dir definition

PhMueller committed:
Merge remote-tracking branch 'origin/FIX93_ContainerConfig' into FIX93_ContainerConfig

PhMueller committed:
Client Abstract Benchmark: Fix bug

PhMueller committed:
ClientAbstract: Add env vars to singularity run command

PhMueller committed:
Update Logging Procedure

PhMueller committed:
Fix container logging

PhMueller committed:
Update Recipe Files:

Add line to clean up unnecessary things which have been created during the installation process.

PhMueller committed:
Fix Container Config Test

PhMueller committed:
Supress some flake warnings

KEggensperger commented on changelog.md:
```diff
@@ -1,5 +1,14 @@
+    The containers had some problems when the file system was read-only. In this case, the home directory, which contains the 
+    hpobenchrc file, was not mounted, so the init of the containers failed. To circumvent this problem, we make multiple
+    changes: 
+    - We introduce an option to mount additional folder. This might be helpful when working on a cluster were the home 
```
were -> where

KEggensperger reviewed (approved):
Could you also update the recipe template?
Other than that looks good and can be merged.

PhMueller committed:
Update Recipe Template + Fix Typo

PhMueller committed:
Merge branch 'development' into FIX93_ContainerConfig
